### PR TITLE
Prune mark distinct columns

### PR DESCRIPTION
**Pull request by alandpost**
Add PruneMarkDistinctColumns and its associated tooling:
* a builder function for MarkDistinctNode
* a matcher for MarkDistinctNode
* a builder function for the matcher
* the rule itself
* a unit test for the rule
* enabling the rule in presto's rule list

The rule elides the MarkDistinctNode entirely if the mark output is not
used.

https://github.com/prestodb/presto/issues/7154

After rebasing after the commit of PruneSemiJoinColumns, the first commit in this PR is now more stand-alone (it used to also contain the first introduction of `Util.restrictOutputs()`)  so I adjusted the commit message.

### CODE REVIEW

**Comment by alandpost** [comment]
Passed DUPA testing:  http://sisyphus.td.teradata.com:8080/job/presto-pipeline-travis/3052/

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java [hunk 1]
```diff
@@ -34,23 +37,22 @@ private Util()
 
     /**
      * Prune the list of available inputs to those required by the given expressions.
-     *
+     * <p>
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: Don't you need closing `</p>` too?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java [hunk 2]
```diff
@@ -34,23 +37,22 @@ private Util()
...
     {
-        Set<Symbol> available = new HashSet<>(availableInputs);
-        Set<Symbol> required = DependencyExtractor.extractUnique(expressions);
-
         // we need to compute the intersection in case some dependencies are symbols from
         // the outer scope (i.e., correlated queries)
-        Set<Symbol> used = Sets.intersection(required, available);
-        if (used.size() == available.size()) {
+        List<Symbol> prunedInputs = availableInputs.stream()
+                .filter(DependencyExtractor.extractUnique(expressions)::contains)
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: I think streams are not that smart to avoid calling `extractUnique` for every symbol. You may want to retrieve it once and pass to the stream-processor.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java [hunk 3]
```diff
@@ -34,23 +37,22 @@ private Util()
...
+                .collect(toImmutableList());
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: You probably don't need to materialize this list in memory upfront. You can postpone materialization till you need the list, e.g.:
```
Stream<Symbol> prunedInputsStream = availableInputs.stream()
        .filter(...);

if (prunedInputsStream.count() == availableInputs.size()) {
  return Optional.empty();
}

return Optional.of(prunedInputsStream.collect(toImmutableList());
...
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java [hunk 4]
```diff
@@ -62,4 +64,23 @@ public static PlanNode transpose(PlanNode parent, PlanNode child)
...
+    {
+        List<Symbol> restrictedOutputs = node.getOutputSymbols().stream()
+                .filter(permittedOutputs::contains)
+                .collect(toImmutableList());
+
+        if (restrictedOutputs.size() < node.getOutputSymbols().size()) {
+            return new ProjectNode(
+                    idAllocator.getNextId(),
+                    node,
+                    Assignments.identity(restrictedOutputs)
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: I think you should keep assignments as in former `PlanNode` .
E.g. in case of `node` being a `ProjectNode` with assignments of  `{ a_16 := a, b_12 := b + 5 }`. For `permittedOutputs` = `{a_16}`, your method would return Assignments of `{ a_16 := a_16 }`.
IMO you should have a special case for `ProjectNode` here (because `ProjectNode` has its own `Assignments`).
Previously it worked fine in context of `CrossJoinElimination` (if there were `ProjectNode` it'd be correct already but in case it's missing, we had to add it) but once you make it general-purpose util function, you should make sure it's correct in all cases.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushProjectionThroughExchange.java [hunk 5]
```diff
@@ -149,10 +150,8 @@
                 newSourceBuilder.build(),
                 inputsBuilder.build());
 
-        if (!result.getOutputSymbols().equals(project.getOutputSymbols())) {
-            // we need to strip unnecessary symbols (hash, partitioning columns).
-            result = new ProjectNode(idAllocator.getNextId(), result, identity(project.getOutputSymbols()));
-        }
+        // we need to strip unnecessary symbols (hash, partitioning columns).
+        result = restrictOutputs(idAllocator, result, ImmutableSet.copyOf(project.getOutputSymbols()));
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushProjectionThroughExchange.java]: `return Optional.of(restrictOutputs(...))`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java (hunk 4 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: You can probably also write a unit test for this method or two to illustrate that problem.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushProjectionThroughExchange.java (hunk 5 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushProjectionThroughExchange.java]: What if `project` had a larger set of output symbols than `result` node? Is such situation legal? Should we add some assertion here? Because `restrictOutputs` clearly allows node having only a subset of `permittedOutputs`, which is not what was checked here before (sets had to be equal).
I'm leaning towards assertion here.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestEliminateCrossJoins.java [hunk 6]
```diff
@@ -65,16 +64,14 @@ public void testEliminateCrossJoin()
                 .setSystemProperty(REORDER_JOINS, "true")
                 .on(crossJoinAndJoin(INNER))
                 .matches(
-                        project(
```
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestEliminateCrossJoins.java]: Can you explain why we don't need `project` here anymore?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java (hunk 4 above)
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: It's not a correctness problem.  The former PlanNode is still there, now as a child.  So the narrower ProjectNode would be the parent of the original PlanNode, in your example.  Rewriting the original ProjectNode is what the output-pruning optimization does.  In combination with RemoveRedundantIdentityProjections, the plan will eventually become what you want.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushProjectionThroughExchange.java (hunk 5 above)
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushProjectionThroughExchange.java]: If project had a larger set of output symbols than result, then the previous code would have produced an invalid identity projection that referred to non-existing symbols.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestEliminateCrossJoins.java (hunk 6 above)
Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestEliminateCrossJoins.java]: The previous code was introducing a project only because the output order had changed, which was not meaningful.  The new plan is better.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneMarkDistinctColumns.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+        Set<Symbol> requiredInputs = Streams.concat(
+                prunedOutputs.get().stream()
+                        .filter(symbol -> !symbol.equals(markDistinctNode.getMarkerSymbol())),
+                markDistinctNode.getDistinctSymbols().stream(),
+                markDistinctNode.getHashSymbol().map(Stream::of).orElse(Stream.empty()))
+                .collect(toImmutableSet());
+
+        return Optional.of(
+                node.replaceChildren(ImmutableList.of(
```
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneMarkDistinctColumns.java]: Sure, will do.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java (hunk 1 above)
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: IntelliJ added that in its automatic reformatting.  Maybe the reformatting rules have changed in an update to the IDE?  In HTML, tags don't need matching closes.  But I'll just delete the blank line, as this whole thing is just a waste of time.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java (hunk 2 above)
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.13.3
```Evaluation of a method reference expression is distinct from invocation of the method itself.

First, if the method reference expression begins with an ExpressionName or a Primary, this subexpression is evaluated.```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java (hunk 3 above)
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: https://docs.oracle.com/javase/8/docs/api/java/util/stream/package-summary.html#StreamOps
"After the terminal operation is performed, the stream pipeline is considered consumed, and can no longer be used"
https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#count--
"[count] is a terminal operation."

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java (hunk 4 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: Makes sense with leaving the `ProjectNode` as a child here.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EliminateCrossJoins.java [hunk 8]
```diff
@@ -191,15 +193,8 @@ public static PlanNode buildJoinTree(List<Symbol> expectedOutputSymbols, JoinGra
...
-            // Introduce a projection to constrain the outputs to what was originally expected
-            // Some nodes are sensitive to what's produced (e.g., DistinctLimit node)
-            result = new ProjectNode(
-                    idAllocator.getNextId(),
-                    result,
-                    Assignments.identity(expectedOutputSymbols));
-        }
-
-        return result;
+        // Introduce a projection to constrain the outputs to what was originally expected
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EliminateCrossJoins.java]: This comment is not true anymore.
It's not always introducing projection.

Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EliminateCrossJoins.java]: Fixed and pushed.

**Comment by maciejgrzybek** [comment]
@martint: seems like ready for you :)

**Comment by alandpost** [comment]
There's a bit more work I want to do on this

**Comment by alandpost** [comment]
Maciej, this is the modification to the pruning function that I was talking about.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneMarkDistinctColumns.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+            return Optional.empty();
+        }
+
+        if (!prunedOutputs.get().contains(markDistinctNode.getMarkerSymbol())) {
+            return Optional.of(
+                    node.replaceChildren(ImmutableList.of(markDistinctNode.getSource())));
+        }
+
+        Set<Symbol> requiredInputs = Streams.concat(
+                prunedOutputs.get().stream()
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneMarkDistinctColumns.java]: I think you need deeper indentation here, see line 68.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneMarkDistinctColumns.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+                        .filter(symbol -> !symbol.equals(markDistinctNode.getMarkerSymbol())),
+                markDistinctNode.getDistinctSymbols().stream(),
+                markDistinctNode.getHashSymbol().map(Stream::of).orElse(Stream.empty()))
+                .collect(toImmutableSet());
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneMarkDistinctColumns.java]: I think you could extract a `DependencyExtractor` which would be able to extract such collections for plan nodes. I have seen that it is typical thing in your PRs with rules migration. 

It could be based on non-recursive visitor and does not need to be fully implemented, it can only return proper collections for things which are used.

What do you think?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java [hunk 11]
```diff
@@ -114,6 +115,11 @@ public ProjectNode project(Assignments assignments, PlanNode source)
         return new ProjectNode(idAllocator.getNextId(), source, assignments);
     }
 
+    public MarkDistinctNode markDistinct(Symbol markerSymbol, List<Symbol> distinctSymbols, Optional<Symbol> hashSymbol, PlanNode source)
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java]: What do you think about having two methods, one with hashSymbol and one without? That way you would need to pass Optional (at least into the API of `PlanBuilder`)

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneMarkDistinctColumns.java [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+                                    p.values(key, hash, unused)));
+                })
+                .matches(
+                        strictProject(
+                                ImmutableMap.of("mark", PlanMatchPattern.expression("mark")),
+                                markDistinct("mark", ImmutableList.of("key"), Optional.of("hash"),
+                                        strictProject(
+                                                ImmutableMap.of(
+                                                        "key", PlanMatchPattern.expression("key"),
+                                                        "hash", PlanMatchPattern.expression("hash")),
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneMarkDistinctColumns.java]: static import of `expression` ?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java [hunk 13]
```diff
@@ -163,6 +164,18 @@ public static PlanMatchPattern aggregation(
         return result;
     }
 
+    public static PlanMatchPattern markDistinct(
+            String markerSymbol,
+            List<String> distinctSymbols,
+            Optional<String> hashSymbol,
+            PlanMatchPattern source)
+    {
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java]: What do you think about having two methods, one with hashSymbol and one without? That way you would need to pass Optional (at least into the API of `PlanMatchPattern`).

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushProjectionThroughExchange.java [hunk 14]
```diff
@@ -158,12 +159,8 @@ public Pattern getPattern()
...
                 inputsBuilder.build());
 
-        if (!result.getOutputSymbols().equals(project.getOutputSymbols())) {
-            // we need to strip unnecessary symbols (hash, partitioning columns).
-            result = new ProjectNode(idAllocator.getNextId(), result, identity(project.getOutputSymbols()));
-        }
-
-        return Optional.of(result);
+        // we need to strip unnecessary symbols (hash, partitioning columns).
+        return Optional.of(restrictOutputs(idAllocator, result, ImmutableSet.copyOf(project.getOutputSymbols())).orElse(result));
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushProjectionThroughExchange.java]: do you need `orElse` here?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java [hunk 15]
```diff
@@ -39,23 +37,21 @@ private Util()
...
     public static Optional<List<Symbol>> pruneInputs(Collection<Symbol> availableInputs, Collection<Expression> expressions)
     {
-        Set<Symbol> available = new HashSet<>(availableInputs);
-        Set<Symbol> required = DependencyExtractor.extractUnique(expressions);
-
         // we need to compute the intersection in case some dependencies are symbols from
         // the outer scope (i.e., correlated queries)
-        Set<Symbol> used = Sets.intersection(required, available);
-        if (used.size() == available.size()) {
+        List<Symbol> prunedInputs = availableInputs.stream()
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: Please move changes in `pruneInputs` to separate commit

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestEliminateCrossJoins.java [hunk 16]
```diff
@@ -65,16 +64,14 @@ public void testEliminateCrossJoin()
...
+                        join(INNER,
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestEliminateCrossJoins.java]: Can you please extract changes in this file to separate commit, as it is not related either to changes in `pruneInputs` or `restrictOutputs`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneMarkDistinctColumns.java (hunk 9 above)
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneMarkDistinctColumns.java]: The file is indented according to the IntelliJ rules.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneMarkDistinctColumns.java (hunk 10 above)
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneMarkDistinctColumns.java]: The only user of that extractor would be this rewrite, as far as I know.  Such an extractor could be used as part of an integrated, single rule for iterative output pruning, as I was originally working on for this migration:  https://github.com/Teradata/presto/blob/347feaeb6c58e229b6194242084bb0c8662eff07/presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneUnreferencedOutputs.java
but I don't think it makes sense given the current rule factoring.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java (hunk 11 above)
Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java]: Sure, if you like I'll do that.  Ideally, we'd have builders for plan nodes automatically generated from their declarations, along with things like PlanVisitor.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneMarkDistinctColumns.java (hunk 12 above)
Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPruneMarkDistinctColumns.java]: Yes, I'll do that, thanks.  The qualification was left over from a different test where PlanBuilder.expression was being imported.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java (hunk 13 above)
Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java]: Sure, if you like I'll do that.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushProjectionThroughExchange.java (hunk 14 above)
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushProjectionThroughExchange.java]: Yes.  The outer Optional is about whether the PushProjectionThroughExchange fired, which at this point in the code should always be true, and the inner Optional is about whether a project-off was needed.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java (hunk 15 above)
Comment (alandpost) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java]: OK

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestEliminateCrossJoins.java (hunk 16 above)
Comment (alandpost) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestEliminateCrossJoins.java]: It's required by the change to EliminateCrossJoins.  I'll pull those two files out into a separate commit.

**Comment by alandpost** [comment]
@kokosing Please take another look.  Thanks!

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneMarkDistinctColumns.java (hunk 10 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneMarkDistinctColumns.java]: > The only user of that extractor would be this rewrite, as far as I know.

That would be for now. `SymbolsExtractor` (previously `DependencyExtractor`) is used in plenty places and has strange semantic as it extracts symbols recursively where in most places only first level dependencies are needed. So such proper `DependencyExtractor` could slowly replace `SymbolsExtractor`.

That is just my idea. It is very optional comment ;)

Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneMarkDistinctColumns.java]: > Such an extractor could be used as part of an integrated, single rule for iterative output pruning, as I was originally working on for this migration:

Possibly, as you can see there are new possibilities which shows up after such refactor ;)

**Comment by kokosing** [comment]
I rebased, once a build passes I will merge this.
